chronos/streams/chunkstream.nim

```diff
@@ -0,0 +1,383 @@
+#
+#    Chronos Asynchronous Chunked-Encoding Stream
+#             (c) Copyright 2019-Present
+#         Status Research & Development GmbH
+#
+#              Licensed under either of
+#  Apache License, version 2.0, (LICENSE-APACHEv2)
+#              MIT license (LICENSE-MIT)
+import ../asyncloop, ../timer
+import asyncstream, ../transports/stream, ../transports/common
+export asyncstream, stream, timer, common
+
+const
+  ChunkBufferSize = 4096
+  ChunkHeaderSize = 8
+  ChunkDefaultTimeout = 10.seconds
+  CRLF = @[byte(0x0D), byte(0x0A)]
+
+type
+  ChunkedStreamReader* = ref object of AsyncStreamReader
+    timeout: Duration
+
+  ChunkedStreamWriter* = ref object of AsyncStreamWriter
+
+  ChunkedStreamError* = object of Exception
+  ChunkedStreamTimeoutError* = object of ChunkedStreamError
+  ChunkedStreamProtocolError* = object of ChunkedStreamError
+
+proc newTimeoutError(): ref Exception {.inline.} =
+  newException(ChunkedStreamTimeoutError, "Timeout exceeded!")
+
+proc newProtocolError(): ref Exception {.inline.} =
+  newException(ChunkedStreamProtocolError, "Protocol error!")
+
+proc oneOf*[A, B, C](fut1: Future[A], fut2: Future[B],
+                     fut3: Future[C]): Future[void] =
+  ## Wait for completion of one future from list [`fut1`, `fut2`, `fut3`],
+  ## resulting future do not care about futures error, so you need to check
+  ## `fut1`, `fut2`, `fut3` for error manually.
+  var retFuture = newFuture[void]("chunked.oneOf()")
+  proc cb(data: pointer) {.gcsafe.} =
+    var index: int
+    if not retFuture.finished:
+      var fut = cast[FutureBase](data)
+      if cast[pointer](fut1) == data:
+        fut2.removeCallback(cb)
+        fut3.removeCallback(cb)
+      elif cast[pointer](fut2) == data:
+        fut1.removeCallback(cb)
+        fut3.removeCallback(cb)
+      elif cast[pointer](fut3) == data:
+        fut2.removeCallback(cb)
+        fut3.removeCallback(cb)
+      retFuture.complete()
+  fut1.callback = cb
+  fut2.callback = cb
+  fut3.callback = cb
+  return retFuture
+
+proc oneOf*[A, B](fut1: Future[A], fut2: Future[B]): Future[void] =
+  ## Wait for completion of `fut1` or `fut2`, resulting future do not care about
+  ## error, so you need to check `fut1` and `fut2` for error.
+  var retFuture = newFuture[void]("chunked.oneOf()")
+  proc cb(data: pointer) {.gcsafe.} =
+    var index: int
+    if not retFuture.finished:
+      var fut = cast[FutureBase](data)
+      if cast[pointer](fut1) == data:
+        fut2.removeCallback(cb)
+      elif cast[pointer](fut2) == data:
+        fut1.removeCallback(cb)
+      retFuture.complete()
+  fut1.callback = cb
+  fut2.callback = cb
+  return retFuture
+
+proc getChunkSize(buffer: openarray[byte]): uint64 =
+  # We using `uint64` representation, but allow only 2^32 chunk size,
+  # ChunkHeaderSize.
+  for i in 0..<min(len(buffer), ChunkHeaderSize):
+    let ch = buffer[i]
+    if char(ch) in {'0'..'9', 'a'..'f', 'A'..'F'}:
```

Comment:
Why don't we reuse some existing routine for hex parsing?

Reply:
Because every possible hex parsing we have are not suited for some particular use cases. For example `0x` is not allowed in this particular hex parsing and also characters which are not in hex range are not an exception, but end of the number here.